### PR TITLE
[adaptors_task][MKryvonos] implementation

### PR DESCRIPTION
**Pull request by MKryvonos**

### CODE REVIEW

File: adaptors task/student/include/impl.h [hunk 1]
```diff
@@ -11,4 +13,64 @@
  *
  * @return true if expression is valid
  */
-bool isValid(const std::string& source);
+bool isOpenBracket(char ch, std::vector< std::pair <char,char> >& vector){
+    bool isOpen = false;
+    for(int j=0; j< vector.size();++j){
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:18]: Always prefer range for!

File: adaptors task/student/include/impl.h [hunk 2]
```diff
@@ -11,4 +13,64 @@
...
+            isOpen = true;
+    }
+    return isOpen;
+}
+
+bool isClosedBracket(char ch, std::vector< std::pair <char,char> >& vector){
+    bool isOpen = false;
+    for(int j=0; j< vector.size();++j){
+        if(ch == vector[j].second)
+            isOpen = true;
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:29]: break?

File: adaptors task/student/include/impl.h [hunk 3]
```diff
@@ -11,4 +13,64 @@
...
+    }
+    return isOpen;
+}
+
+char correspondingClosedBracket(char ch, std::vector< std::pair <char,char> >& vector){
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:34]: I like your style, when you separate code to the corresponding functions, thanks!

File: adaptors task/student/include/impl.h [hunk 4]
```diff
@@ -11,4 +13,64 @@
...
+    for(int j=0; j< vector.size();++j){
+        if(ch == vector[j].first)
+            return vector[j].second;
+    }
+}
+bool isValid(const std::string& source){
+    std::stack<char> bracketsStack;
+    std::vector< std::pair <char,char> > bracketsVector;
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:42]: Check .cpp file please, it is already implemented

File: adaptors task/student/include/impl.h [hunk 5]
```diff
@@ -11,4 +13,64 @@
...
+
+    bracketsVector.push_back(std::make_pair('(',')'));
+    bracketsVector.push_back(std::make_pair('{','}'));
+    bracketsVector.push_back(std::make_pair('[',']'));
+
+    bool isValid = true;
+    int lenght = source.size();
+    for(int i=0; i<lenght; i++){
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:50]: Use range for

File: adaptors task/student/include/impl.h [hunk 6]
```diff
@@ -11,4 +13,64 @@
...
+                if(bracketsStack.empty()) {
+                    isValid = false;
+                    break;
+                }
+                char hab = bracketsStack.top();
+                char closedBracket = correspondingClosedBracket(hab,bracketsVector);
+                if(closedBracket == currentChar)
+                    bracketsStack.pop();
+                else {
+                    isValid = false;
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:67]: you can return in this place, don't follow one return rule

File: adaptors task/student/include/impl.h [hunk 7]
```diff
@@ -11,4 +13,64 @@
...
+            bracketsStack.push(currentChar);
+        else {
+            if(!isClosedBracket(currentChar, bracketsVector))
+                continue;
+            else {
+                if(bracketsStack.empty()) {
+                    isValid = false;
+                    break;
+                }
+                char hab = bracketsStack.top();
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:62]: use descriptive names

File: adaptors task/student/include/impl.h [hunk 8]
```diff
@@ -11,4 +13,64 @@
...
+                else {
+                    isValid = false;
+                    break;
+                }
+            }
+        }
+    }
+    if(!bracketsStack.empty())
+        isValid= false;
+    return isValid;
```
Comment (smay1613) [inline adaptors task/student/include/impl.h:75]: return bracketsStack.empty() is enough

File: adaptors task/student/include/impl.h (hunk 1 above)
Comment (MKryvonos) [reply adaptors task/student/include/impl.h]: done

File: adaptors task/student/include/impl.h (hunk 2 above)
Comment (MKryvonos) [reply adaptors task/student/include/impl.h]: added a twist to your idea: "return"

File: adaptors task/student/include/impl.h (hunk 4 above)
Comment (MKryvonos) [reply adaptors task/student/include/impl.h]: done. Thanks!

File: adaptors task/student/include/impl.h (hunk 5 above)
Comment (MKryvonos) [reply adaptors task/student/include/impl.h]: done

File: adaptors task/student/include/impl.h (hunk 6 above)
Comment (MKryvonos) [reply adaptors task/student/include/impl.h]: done

File: adaptors task/student/include/impl.h (hunk 7 above)
Comment (MKryvonos) [reply adaptors task/student/include/impl.h]: renamed

File: adaptors task/student/include/impl.h (hunk 8 above)
Comment (MKryvonos) [reply adaptors task/student/include/impl.h]: done

**Pull request closed without merging**